From 81dc31886af7d79d691b946d4a21c69db4b1d066 Mon Sep 17 00:00:00 2001
From: Gatien Chevallier <gatien.chevallier@foss.st.com>
Date: Wed, 24 Jul 2024 18:00:17 +0200
Subject: [PATCH 1048/1695] drivers: stm32_ipcc: support non-TDCID probe and
 RIF config application

When a CID filtering is previously applied, the IPCC_CxSECCFGR and
IPCC_CxPRIVCFGR registers can only be written by the authorized CID.

In case we're booting as non-TDCID, remove the constraint on the presence
of the RIF configuration.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git a2d39b0488b331cc66dd7063324a320f8525a0e2 drivers: stm32_ipcc: support non-TDCID probe and RIF config application]
Change-Id: I424ca8ace98d8f191564792c0bb0464ba34b07bf
Signed-off-by: Gatien Chevallier <gatien.chevallier@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/395780
Tested-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
---
 core/drivers/stm32_ipcc.c | 129 +++++++++++++++++++++++---------------
 1 file changed, 78 insertions(+), 51 deletions(-)

diff --git a/core/drivers/stm32_ipcc.c b/core/drivers/stm32_ipcc.c
index b49899d2e..ca33169de 100644
--- a/core/drivers/stm32_ipcc.c
+++ b/core/drivers/stm32_ipcc.c
@@ -73,6 +73,7 @@
 /*
  * CIDCFGR register bitfields
  */
+#define IPCC_CIDCFGR_CFEN		BIT(0)
 #define IPCC_CIDCFGR_SCID_MASK		GENMASK_32(6, 4)
 #define IPCC_CIDCFGR_CONF_MASK		(_CIDCFGR_CFEN |	 \
 					 IPCC_CIDCFGR_SCID_MASK)
@@ -126,7 +127,8 @@ struct ipcc_pdata {
 	unsigned int nb_channels_cfg;
 	struct clk *ipcc_clock;
 	vaddr_t base;
-	struct rif_conf_data conf_data;
+	struct rif_conf_data *conf_data;
+	bool is_tdcid;
 
 	/* Remote proc base address */
 	vaddr_t rbase;
@@ -144,15 +146,22 @@ struct ipcc_pdata {
 static STAILQ_HEAD(, ipcc_pdata) ipcc_list =
 		STAILQ_HEAD_INITIALIZER(ipcc_list);
 
-static void apply_rif_config(struct ipcc_pdata *ipcc_d, bool is_tdcid)
+static void apply_rif_config(struct ipcc_pdata *ipcc_d)
 {
+	bool proc1_cidfilt = false;
+	bool proc2_cidfilt = false;
 	uint32_t priv_proc_1 = 0;
 	uint32_t priv_proc_2 = 0;
 	uint32_t sec_proc_1 = 0;
 	uint32_t sec_proc_2 = 0;
+	unsigned int cid1 = 0;
+	unsigned int cid2 = 0;
 	unsigned int i = 0;
 	bool is_cid_configured = false;
 
+	if (!ipcc_d->conf_data)
+		return;
+
 	/*
 	 * Check that the number of channel supported by hardware
 	 * is coherent with the config
@@ -166,7 +175,7 @@ static void apply_rif_config(struct ipcc_pdata *ipcc_d, bool is_tdcid)
 	 * configuration. Clearing previous configuration prevents
 	 * undesired events during the only legitimate configuration.
 	 */
-	if (is_tdcid) {
+	if (ipcc_d->is_tdcid) {
 		/* IPCC Processor 1 */
 		io_clrbits32(ipcc_d->base + IPCC_C1CIDCFGR,
 			     IPCC_CIDCFGR_CONF_MASK);
@@ -176,30 +185,44 @@ static void apply_rif_config(struct ipcc_pdata *ipcc_d, bool is_tdcid)
 			     IPCC_CIDCFGR_CONF_MASK);
 	}
 
+	cid1 = io_read32(ipcc_d->base + IPCC_C1CIDCFGR) &
+	       IPCC_CIDCFGR_SCID_MASK;
+	cid2 = io_read32(ipcc_d->base + IPCC_C2CIDCFGR) &
+	       IPCC_CIDCFGR_SCID_MASK;
+
+	proc1_cidfilt = io_read32(ipcc_d->base + IPCC_C1CIDCFGR) &
+		       IPCC_CIDCFGR_CFEN;
+	proc2_cidfilt = io_read32(ipcc_d->base + IPCC_C2CIDCFGR) &
+		       IPCC_CIDCFGR_CFEN;
+
 	/* Split the sec and priv configuration for IPCC processor 1 and 2 */
-	sec_proc_1 = ipcc_d->conf_data.sec_conf[0] &
-		     GENMASK_32(IPCC_NB_MAX_RIF_CHAN - 1, 0);
-	priv_proc_1 = ipcc_d->conf_data.priv_conf[0] &
+	sec_proc_1 = ipcc_d->conf_data->sec_conf[0] &
 		     GENMASK_32(IPCC_NB_MAX_RIF_CHAN - 1, 0);
+	priv_proc_1 = ipcc_d->conf_data->priv_conf[0] &
+		      GENMASK_32(IPCC_NB_MAX_RIF_CHAN - 1, 0);
 
-	sec_proc_2 = (ipcc_d->conf_data.sec_conf[0] &
+	sec_proc_2 = (ipcc_d->conf_data->sec_conf[0] &
 		      GENMASK_32((IPCC_NB_MAX_RIF_CHAN * 2) - 1,
 				 IPCC_NB_MAX_RIF_CHAN)) >>
 		     IPCC_NB_MAX_RIF_CHAN;
-	priv_proc_2 = (ipcc_d->conf_data.priv_conf[0] &
+	priv_proc_2 = (ipcc_d->conf_data->priv_conf[0] &
 		       GENMASK_32((IPCC_NB_MAX_RIF_CHAN * 2) - 1,
 				  IPCC_NB_MAX_RIF_CHAN)) >>
 		      IPCC_NB_MAX_RIF_CHAN;
 
 	/* Security and privilege RIF configuration */
-	io_clrsetbits32(ipcc_d->base + IPCC_C1PRIVCFGR, IPCC_PRIVCFGR_MASK,
-			priv_proc_1);
-	io_clrsetbits32(ipcc_d->base + IPCC_C2PRIVCFGR, IPCC_PRIVCFGR_MASK,
-			priv_proc_2);
-	io_clrsetbits32(ipcc_d->base + IPCC_C1SECCFGR, IPCC_SECCFGR_MASK,
-			sec_proc_1);
-	io_clrsetbits32(ipcc_d->base + IPCC_C2SECCFGR, IPCC_SECCFGR_MASK,
-			sec_proc_2);
+	if (!proc1_cidfilt || cid1 == RIF_CID1) {
+		io_clrsetbits32(ipcc_d->base + IPCC_C1PRIVCFGR,
+				IPCC_PRIVCFGR_MASK, priv_proc_1);
+		io_clrsetbits32(ipcc_d->base + IPCC_C1SECCFGR,
+				IPCC_SECCFGR_MASK, sec_proc_1);
+	}
+	if (!proc2_cidfilt || cid2 == RIF_CID1) {
+		io_clrsetbits32(ipcc_d->base + IPCC_C2PRIVCFGR,
+				IPCC_PRIVCFGR_MASK, priv_proc_2);
+		io_clrsetbits32(ipcc_d->base + IPCC_C2SECCFGR,
+				IPCC_SECCFGR_MASK, sec_proc_2);
+	}
 
 	/*
 	 * Evaluate RIF CID filtering configuration before setting it.
@@ -208,28 +231,28 @@ static void apply_rif_config(struct ipcc_pdata *ipcc_d, bool is_tdcid)
 	 * channels of this processor. This is a configuration check.
 	 */
 	for (i = 0; i < IPCC_NB_MAX_RIF_CHAN; i++) {
-		if (!(BIT(i) & ipcc_d->conf_data.access_mask[0]))
+		if (!(BIT(i) & ipcc_d->conf_data->access_mask[0]))
 			continue;
 
 		if (!is_cid_configured &&
-		    (BIT(0) & ipcc_d->conf_data.cid_confs[i])) {
+		    (BIT(0) & ipcc_d->conf_data->cid_confs[i])) {
 			is_cid_configured = true;
 			if (i == IPCC_NB_MAX_RIF_CHAN - 1)
 				panic("Inconsistent IPCC CID filtering RIF configuration");
 		}
 
 		if (is_cid_configured &&
-		    !(BIT(0) & ipcc_d->conf_data.cid_confs[i]))
+		    !(BIT(0) & ipcc_d->conf_data->cid_confs[i]))
 			panic("Inconsistent IPCC CID filtering RIF configuration");
 	}
 
 	/* IPCC processor 1 CID filtering configuration */
-	if (!is_tdcid)
+	if (!ipcc_d->is_tdcid)
 		return;
 
 	io_clrsetbits32(ipcc_d->base + IPCC_C1CIDCFGR,
 			IPCC_CIDCFGR_CONF_MASK,
-			ipcc_d->conf_data.cid_confs[0]);
+			ipcc_d->conf_data->cid_confs[0]);
 
 	/*
 	 * Reset this field to evaluate CID filtering configuration
@@ -238,30 +261,30 @@ static void apply_rif_config(struct ipcc_pdata *ipcc_d, bool is_tdcid)
 	is_cid_configured = false;
 
 	for (i = IPCC_NB_MAX_RIF_CHAN; i < IPCC_NB_MAX_RIF_CHAN * 2; i++) {
-		if (!(BIT(i) & ipcc_d->conf_data.access_mask[0]))
+		if (!(BIT(i) & ipcc_d->conf_data->access_mask[0]))
 			continue;
 
 		if (!is_cid_configured &&
-		    (BIT(0) & ipcc_d->conf_data.cid_confs[i])) {
+		    (BIT(0) & ipcc_d->conf_data->cid_confs[i])) {
 			is_cid_configured = true;
 			if (i == (IPCC_NB_MAX_RIF_CHAN * 2) - 1)
 				panic("Inconsistent IPCC CID filtering RIF configuration");
 		}
 
 		if (is_cid_configured &&
-		    !(BIT(0) & ipcc_d->conf_data.cid_confs[i]))
+		    !(BIT(0) & ipcc_d->conf_data->cid_confs[i]))
 			panic("Inconsistent IPCC CID filtering RIF configuration");
 	}
 
 	/* IPCC Processor 2 CID filtering configuration */
 	io_clrsetbits32(ipcc_d->base + IPCC_C2CIDCFGR,
 			IPCC_CIDCFGR_CONF_MASK,
-			ipcc_d->conf_data.cid_confs[IPCC_NB_MAX_RIF_CHAN]);
+			ipcc_d->conf_data->cid_confs[IPCC_NB_MAX_RIF_CHAN]);
 }
 
 static void stm32_ipcc_pm_resume(struct ipcc_pdata *ipcc)
 {
-	apply_rif_config(ipcc, true);
+	apply_rif_config(ipcc);
 }
 
 static void stm32_ipcc_pm_suspend(struct ipcc_pdata *ipcc __unused)
@@ -279,12 +302,8 @@ stm32_ipcc_pm(enum pm_op op, unsigned int pm_hint,
 {
 	struct ipcc_pdata *ipcc = pm_handle->handle;
 	TEE_Result res = TEE_ERROR_GENERIC;
-	bool is_tdcid = false;
 
-	if (stm32_rifsc_check_tdcid(&is_tdcid))
-		panic();
-
-	if (!PM_HINT_IS_STATE(pm_hint, CONTEXT) || !is_tdcid)
+	if (!PM_HINT_IS_STATE(pm_hint, CONTEXT) || !ipcc->is_tdcid)
 		return TEE_SUCCESS;
 
 	res = clk_enable(ipcc->ipcc_clock);
@@ -670,25 +689,31 @@ static TEE_Result parse_dt(const void *fdt, int node, struct ipcc_pdata *ipcc_d)
 		return res;
 
 	cuint = fdt_getprop(fdt, node, "st,protreg", &lenp);
-	if (!cuint)
-		panic("No RIF configuration available");
+	if (!cuint) {
+		DMSG("No RIF configuration available");
+		return TEE_SUCCESS;
+	}
+
+	ipcc_d->conf_data = calloc(1, sizeof(*ipcc_d->conf_data));
+	if (!ipcc_d->conf_data)
+		panic();
 
 	ipcc_d->nb_channels_cfg = (unsigned int)(lenp / sizeof(uint32_t));
 	assert(ipcc_d->nb_channels_cfg <= (IPCC_NB_MAX_RIF_CHAN * 2));
 
-	ipcc_d->conf_data.cid_confs = calloc(IPCC_NB_MAX_RIF_CHAN * 2,
-					     sizeof(uint32_t));
-	ipcc_d->conf_data.sec_conf = calloc(1, sizeof(uint32_t));
-	ipcc_d->conf_data.priv_conf = calloc(1, sizeof(uint32_t));
-	ipcc_d->conf_data.access_mask = calloc(1, sizeof(uint32_t));
-	if (!ipcc_d->conf_data.cid_confs || !ipcc_d->conf_data.sec_conf ||
-	    !ipcc_d->conf_data.priv_conf || !ipcc_d->conf_data.access_mask)
+	ipcc_d->conf_data->cid_confs = calloc(IPCC_NB_MAX_RIF_CHAN * 2,
+					      sizeof(uint32_t));
+	ipcc_d->conf_data->sec_conf = calloc(1, sizeof(uint32_t));
+	ipcc_d->conf_data->priv_conf = calloc(1, sizeof(uint32_t));
+	ipcc_d->conf_data->access_mask = calloc(1, sizeof(uint32_t));
+	if (!ipcc_d->conf_data->cid_confs || !ipcc_d->conf_data->sec_conf ||
+	    !ipcc_d->conf_data->priv_conf || !ipcc_d->conf_data->access_mask)
 		panic("Missing memory capacity for ipcc RIF configuration");
 
 	for (i = 0; i < ipcc_d->nb_channels_cfg; i++) {
 		rif_conf = fdt32_to_cpu(cuint[i]);
 
-		stm32_rif_parse_cfg(rif_conf, &ipcc_d->conf_data,
+		stm32_rif_parse_cfg(rif_conf, ipcc_d->conf_data,
 				    IPCC_NB_MAX_CID_SUPPORTED,
 				    IPCC_NB_MAX_RIF_CHAN * 2);
 	}
@@ -701,16 +726,15 @@ static TEE_Result stm32_ipcc_probe(const void *fdt, int node,
 {
 	TEE_Result res = TEE_ERROR_GENERIC;
 	struct ipcc_pdata *ipcc_d = NULL;
-	bool is_tdcid = false;
-
-	res = stm32_rifsc_check_tdcid(&is_tdcid);
-	if (res)
-		return res;
 
 	ipcc_d = calloc(1, sizeof(*ipcc_d));
 	if (!ipcc_d)
 		return TEE_ERROR_OUT_OF_MEMORY;
 
+	res = stm32_rifsc_check_tdcid(&ipcc_d->is_tdcid);
+	if (res)
+		goto err;
+
 	res = parse_dt(fdt, node, ipcc_d);
 	if (res)
 		goto err;
@@ -719,7 +743,7 @@ static TEE_Result stm32_ipcc_probe(const void *fdt, int node,
 	if (res)
 		panic("Cannot access IPCC clock");
 
-	apply_rif_config(ipcc_d, is_tdcid);
+	apply_rif_config(ipcc_d);
 
 	clk_disable(ipcc_d->ipcc_clock);
 
@@ -734,10 +758,13 @@ static TEE_Result stm32_ipcc_probe(const void *fdt, int node,
 
 err:
 	/* Free all allocated resources */
-	free(ipcc_d->conf_data.access_mask);
-	free(ipcc_d->conf_data.cid_confs);
-	free(ipcc_d->conf_data.priv_conf);
-	free(ipcc_d->conf_data.sec_conf);
+	if (ipcc_d->conf_data) {
+		free(ipcc_d->conf_data->access_mask);
+		free(ipcc_d->conf_data->cid_confs);
+		free(ipcc_d->conf_data->priv_conf);
+		free(ipcc_d->conf_data->sec_conf);
+	}
+	free(ipcc_d->conf_data);
 	free(ipcc_d);
 
 	return res;
-- 
2.47.2

